lib/core/src/Cardano/Wallet/DB/Sqlite/CheckpointsOld.hs

```diff
@@ -396,67 +392,48 @@ instance PersistAddressBook (Seq.SeqState n k)
     loadDiscoveries wid sl = DS <$> loadDiscoveries wid sl
 
 instance
-    ( Eq (key 'AccountK XPub)
-    , PersistPublicKey (key 'AccountK)
+    ( PersistPublicKey (key 'AccountK)
     , PersistPublicKey (key 'AddressK)
     , MkKeyFingerprint key (Proxy n, key 'AddressK XPub)
-    , GetPurpose key
     , PaymentAddress n key
     , SoftDerivation key
+    , Typeable n
     , (key == SharedKey) ~ 'False
     ) => PersistAddressBook (Seq.SeqState n key) where
 
     insertPrologue wid (SeqPrologue st) = do
-        let (intPool, extPool) =
-                (Seq.internalPool st, Seq.externalPool st)
-        let (Seq.ParentContextUtxo accXPubInternal) = Seq.context intPool
-        let (Seq.ParentContextUtxo accXPubExternal) = Seq.context extPool
-        let (accountXPub, _) = invariant
-                "Internal & External pool use different account public keys!"
-                ( accXPubExternal, accXPubInternal )
-                (uncurry (==))
-        let eGap = Seq.gap extPool
-        let iGap = Seq.gap intPool
         repsert (SeqStateKey wid) $ SeqState
             { seqStateWalletId = wid
-            , seqStateExternalGap = eGap
-            , seqStateInternalGap = iGap
-            , seqStateAccountXPub = serializeXPub accountXPub
-            , seqStateRewardXPub = serializeXPub (Seq.rewardAccountKey st)
+            , seqStateExternalGap = Seq.getGap $ Seq.externalPool st
+            , seqStateInternalGap = Seq.getGap $ Seq.internalPool st
+            , seqStateAccountXPub = serializeXPub $ Seq.accountXPub st
+            , seqStateRewardXPub = serializeXPub $ Seq.rewardAccountKey st
             , seqStateDerivationPrefix = Seq.derivationPrefix st
             }
         deleteWhere [SeqStatePendingWalletId ==. wid]
         dbChunked
             insertMany_
             (mkSeqStatePendingIxs wid $ Seq.pendingChangeIxs st)
 
-    insertDiscoveries wid sl
-        (SeqDiscoveries (SeqAddressList ints) (SeqAddressList exts))
-      = do
-        void $ dbChunked insertMany_
-            [ SeqStateAddress wid sl addr ix UtxoInternal state
-            | (ix, (addr, state)) <- zip [0..] ints
-            ]
-        void $ dbChunked insertMany_
-            [ SeqStateAddress wid sl addr ix UtxoExternal state
-            | (ix, (addr, state)) <- zip [0..] exts
-            ]
+    insertDiscoveries wid sl (SeqDiscoveries ints exts) = do
+        insertSeqAddressMap @n wid sl ints
```

Comment:
indeed, the erased two structures begged for reuse. UtxoInternal and UntixExternal was the only difference between them

Reply:
I didn't mind the reuse; the thing I like most about the new formulation (which was there before at some point) is that it ensures that e.g. `int` is paired with `UtxoInternal`. In contrast, the structures removed here would, in principle, permit switching `ints` and `exts` without type error.